File: Lab_RGR/Proxy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_RGR
{
    // Structural Pattern - Proxy
    public interface IPhotoViewer
    {
        void DisplayPhoto();
    }

    public class PhotoViewer : IPhotoViewer
    {
        private string _photo;

        public PhotoViewer(string photo)
        {
            _photo = photo;
        }

        public void DisplayPhoto()
        {
            Console.WriteLine($"Displaying photo: {_photo}");
        }
    }

    public class ProxyPhotoViewer : IPhotoViewer
    {
        private readonly string _photo;
        private PhotoViewer _photoViewer;

        public ProxyPhotoViewer(string photo)
        {
            _photo = photo;
        }

        public void DisplayPhoto()
        {
            if (_photoViewer == null)
            {
                _photoViewer = new PhotoViewer(_photo);
            }
            _photoViewer.DisplayPhoto();
        }
    }

}

File: Python_Patterns/Proxy.py
class IPhotoViewer:
    def display_photo(self):
        pass


class PhotoViewer(IPhotoViewer):
    def __init__(self, photo):
        self._photo = photo

    def display_photo(self):
        print(f"Displaying photo: {self._photo}")


class ProxyPhotoViewer(IPhotoViewer):
    def __init__(self, photo):
        self._photo = photo
        self._photo_viewer = None

    def display_photo(self):
        if self._photo_viewer is None:
            self._photo_viewer = PhotoViewer(self._photo)
        self._photo_viewer.display_photo()
